### DISCUSSION TITLE
TuyaMCU not communicating after init sequence on smoke detector with ESP12F replacement for CB3S

### DISCUSSION THREAD

**Discussion by belveder79** [arendst/Tasmota#18935] [Q&A]
### PROBLEM DESCRIPTION

I got a smoke detector from AliExpress and replaced the CB3S module with an ESP12F module with stock tasmota firmware on it. The problem I'm facing is that it initializes right as you can see from the log
```lua
MCU Product ID: {"p":"fzb5up2iehjcrnqy","v":"1.0.0"}
```
but from there on nothing happens or works. The entire assembly is admittedly supposed to be super low power and I power the ESP externally to not go to sleep straight after the initialization, but I'd expect the system to be active at least for a couple of seconds (not just 2-3 or so). Anyway, the overall issue is that as far as I can see the MCU does not respond straight after the product ID. What is suspicious to me is this:

```lua
14:07:27.094 TYA: Send "55aa0002000001"
14:07:27.102 {"TuyaReceived":{"Data":"55AA0002000001","Cmnd":2}}
14:07:27.103 TYA: RX MCU configuration Mode=0
14:07:27.104 TYA: Read MCU state
14:07:27.112 TYA: Send "55aa0008000007"
```

The read from TuyaReceived is the same as the Send. I looked into the Tuya Serial Documentation but it did not really shed some insight (it should return MCU GPIO config, but it apparently does not). 

I also tried to run with the V2 version of the Tuya driver with a custom tasmota build, but it did not work at all...

Any hint would be helpful (going to sniff the serial on the original CB3S chip now to see what is communicated).

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): TuyaMCU
- [x] Tasmota binary firmware version number used: 12.5.0.4
  - [x] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: web/esptool/ota
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"2272":"Tuya Tx"},"GPIO2":{"0":"None"},"GPIO3":{"2304":"Tuya Rx"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
No rules defined...

14:04:01.851 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 19, Pld -99, Data 'Rule1; Rule2; Rule3'
14:04:01.867 SRC: Backlog
14:04:01.869 CMD: Grp 0, Cmd 'RULE', Idx 1, Len 0, Pld -99, Data ''
14:04:01.874 RSL: RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
14:04:02.093 SRC: Backlog
14:04:02.096 CMD: Grp 0, Cmd 'RULE', Idx 2, Len 0, Pld -99, Data ''
14:04:02.101 RSL: RESULT = {"Rule2":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
14:04:02.345 SRC: Backlog
14:04:02.347 CMD: Grp 0, Cmd 'RULE', Idx 3, Len 0, Pld -99, Data ''
14:04:02.352 RSL: RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
```
- [ ] Provide the output of this command: `Status 0`:
```lua
14:04:35.842 CMD: Grp 0, Cmd 'STATUS', Idx 1, Len 1, Pld 0, Data '0'
14:04:35.847 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"firealarm","FriendlyName":["firealarm2"],"Topic":"tasmota_D8CC16","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
14:04:35.853 RSL: STATUS1 = {"StatusPRM":{"Baudrate":9600,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"External System","Uptime":"0T00:02:02","StartupUTC":"2023-06-22T13:02:33","Sleep":50,"CfgHolder":4617,"BootCount":86,"BCResetTime":"2022-10-03T11:23:35","SaveCount":135,"SaveAddress":"F9000"}}
14:04:35.856 RSL: STATUS2 = {"StatusFWR":{"Version":"12.5.0.4(tasmota)","BuildDateTime":"2023-06-21T16:34:43","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"345/699"}}
14:04:35.860 RSL: STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":4,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["NETGEAR81",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
14:04:35.871 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":634,"Free":368,"Heap":24,"ProgramFlashSize":1024,"FlashSize":4096,"FlashChipId":"164068","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000080"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62,68","Sensors":"1,2,3,4,5,6","I2CDriver":"7"}}
14:04:35.877 RSL: STATUS5 = {"StatusNET":{"Hostname":"firealarm","IPAddress":"10.0.0.41","Gateway":"10.0.0.1","Subnetmask":"255.255.255.0","DNSServer1":"10.0.0.1","DNSServer2":"0.0.0.0","Mac":"AC:0B:FB:D8:CC:16","Webserver":2,"HTTP_API":1,"WifiConfig":2,"WifiPower":17.0}}
14:04:35.881 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_D8CC16","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
14:04:35.887 RSL: STATUS7 = {"StatusTIM":{"UTC":"2023-06-22T13:04:35","Local":"2023-06-22T14:04:35","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"04:46","Sunset":"20:57"}}
14:04:35.893 RSL: STATUS10 = {"StatusSNS":{"Time":"2023-06-22T14:04:35"}}
14:04:35.898 RSL: STATUS11 = {"StatusSTS":{"Time":"2023-06-22T14:04:35","Uptime":"0T00:02:02","UptimeSec":122,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"NETGEAR81","BSSId":"10:DA:43:84:25:B5","Channel":4,"Mode":"11n","RSSI":74,"Signal":-63,"LinkCount":1,"Downtime":"0T00:00:03"}}}
```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
00:00:00.075 TYA: Request MCU configuration at 9600 bps
00:00:00.076 SRC: Restart
00:00:00.089 TYA: Send "55aa0006000501010001000d"
00:00:00.111 Project tasmota - firealarm Version 12.5.0.4(tasmota)-2_7_4_9(2023-06-21T16:34:43)
00:00:00.213 WIF: Checking connection...
00:00:00.213 WIF: Attempting connection...
00:00:00.550 WIF: Connecting to AP1 NETGEAR81 Channel 4 BSSId 10:DA:43:84:25:B5 in mode 11n as firealarm...
00:00:01.003 TYA: Set WiFi LED 2 (-1)
00:00:01.012 TYA: Send "55aa000300010205"
00:00:01.754 WIF: Checking connection...
00:00:01.755 WIF: Connected
00:00:02.006 HTP: Web server active on firealarm with IP address 10.0.0.41
00:00:02.403 WIF: Sending Gratuitous ARP
00:00:02.404 NTP: Sync time...
00:00:02.411 WIF: DNS resolved 'pool.ntp.org' (185.119.117.217) in 5 ms
00:00:02.435 RTC: UTC 2023-06-22T13:07:26, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
14:07:26.000 RTC: Synced by NTP
14:07:26.002 TYA: Set WiFi LED 3 (0)
14:07:26.011 TYA: Send "55aa000300010306"
14:07:26.275 HTP: Console
14:07:27.021 RSL: INFO1 = {"Info1":{"Module":"WifiSmoke","Version":"12.5.0.4(tasmota)","FallbackTopic":"cmnd/DVES_D8CC16_fb/","GroupTopic":"cmnd/tasmotas/"}}
14:07:27.023 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"firealarm","IPAddress":"10.0.0.41"}}
14:07:27.025 RSL: INFO3 = {"Info3":{"RestartReason":"External System","BootCount":87}}
14:07:27.027 RSL: RESULT = {"POWER":"OFF"}
14:07:27.029 RSL: POWER = OFF
14:07:27.037 TYA: Send "55aa0001000000"
14:07:27.084 {"TuyaReceived":{"Data":"55AA000100247B2270223A22667A62357570326965686A63726E7179222C2276223A22312E302E30227DFA","Cmnd":1,"CmndData":"7B2270223A22667A62357570326965686A63726E7179222C2276223A22312E302E30227D"}}
14:07:27.086 TYA: MCU Product ID: {"p":"fzb5up2iehjcrnqy","v":"1.0.0"}
14:07:27.094 TYA: Send "55aa0002000001"
14:07:27.102 {"TuyaReceived":{"Data":"55AA0002000001","Cmnd":2}}
14:07:27.103 TYA: RX MCU configuration Mode=0
14:07:27.104 TYA: Read MCU state
14:07:27.112 TYA: Send "55aa0008000007"
14:07:30.464 QPC: Reset
14:07:31.428 RSL: STATE = {"Time":"2023-06-22T14:07:31","Uptime":"0T00:00:09","UptimeSec":9,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"NETGEAR81","BSSId":"10:DA:43:84:25:B5","Channel":4,"Mode":"11n","RSSI":56,"Signal":-72,"LinkCount":1,"Downtime":"0T00:00:03"}}
14:07:32.426 APP: Boot Count 87
14:07:32.594 CFG: Saved to flash at F8, Count 136, Bytes 4096
14:07:34.411 TYA: Send "55aa00000000ff"
14:07:36.424 RSL: config = {"ip":"10.0.0.41","dn":"firealarm","fn":["firealarm2",null,null,null,null,null,null,null],"hn":"firealarm","mac":"AC0BFBD8CC16","md":"WifiSmoke","ty":1,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"12.5.0.4","t":"tasmota_D8CC16","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"sho":[],"sht":[],"ver":1}
14:07:36.428 RSL: sensors = {"sn":{"Time":"2023-06-22T14:07:36"},"ver":1}
14:07:45.392 TYA: Send "55aa00000000ff"
14:07:45.787 WIF: Checking connection...
14:07:56.413 TYA: Send "55aa00000000ff"
14:08:05.795 WIF: Checking connection...
14:08:07.408 TYA: Send "55aa00000000ff"
14:08:18.396 TYA: Send "55aa00000000ff"
14:08:22.418 TYA: Send "55aa001c0008011706160e08160487"
14:08:24.408 WIF: Sending Gratuitous ARP
14:08:25.759 WIF: Checking connection...
14:08:29.433 TYA: Send "55aa00000000ff"
14:08:40.411 TYA: Send "55aa00000000ff"
14:08:45.774 WIF: Checking connection...
14:08:51.396 TYA: Send "55aa00000000ff"
14:08:51.847 HTP: Console```

### TO REPRODUCE

1) Buy https://www.aliexpress.com/item/1005005480282632.html?spm=a2g0o.order_list.order_list_main.5.6da01802l7Wht1
2) replace CB3S with ESP12 and tasmota, 
3) try...


### EXPECTED BEHAVIOUR

MCU communicating stuff back to tasmota, or tasmota running some init sequence to the MCU?


### SCREENSHOTS

![IMG_7620](https://github.com/arendst/Tasmota/assets/112880/b2942f1b-66d3-47df-b73b-c196256410d8)
![IMG_7622](https://github.com/arendst/Tasmota/assets/112880/c221a0c4-4d3e-4e3e-9cc6-4f0a90f801ea)
![IMG_7621](https://github.com/arendst/Tasmota/assets/112880/dd17ac72-dfd2-45c2-85a8-025e7460242b)

### ADDITIONAL CONTEXT

I know that the Tuya protocol is a work in progress and I'm happy to contribute to that if necessary. It seems that the current status is partially working for this MCU at least, but I couldn't figure out what's wrong currently...

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment]
> The read from TuyaReceived is the same as the Send.
That could be suspicious and you'd better check for an permanent or intermittent contact between TX and RX.

However other logs doesn't seem to confirm
```
14:07:27.037 TYA: Send "55aa0001000000"
14:07:27.084 {"TuyaReceived":{"Data":"55AA000100247B2270223A22667A62357570326965686A63726E7179222C2276223A22312E302E30227DFA","Cmnd":1,"CmndData":"7B2270223A22667A62357570326965686A63726E7179222C2276223A22312E302E30227D"}}
```

Is this device documented in the Template repository ?
If not you may be missing some configuration
What is it supposed to do ?
Have you tried to set smoke on it to see if that triggers some messages ?

**Comment by sfromis** [comment]
With a smoke detector, I'd expect battery operation, which normally means that the TuyaMCU chip as the main processor only expects the Wifi chip to be woken up when there is a need for Wifi. Existing smoke detector configs do have quite a bit more than just configuring the pins needed. Here's an example (which may not at all apply to your device).
https://templates.blakadder.com/YG400A.html

**Comment by belveder79** [comment]
@barbudor actually the device is not documented yet actually. I will move on with checking on the connections between the ESP side and the MCU side. I did not put something on flames to test it, no :-) but there is a "WIFI enable" button and a "Test" button that really beeps if you press it and the one supplies power to the ESP again, which makes total sense to me. However, there is nothing happening in the Serial or nothing received if I do either of that...

@sfromis indeed the MCU puts the ESP to sleep about 30 seconds after initial power up, while "putting to sleep" refers to "do not provide any power to the ESP any more". In this respect the TuyaMCU behaves exactly the same as e.g. this one I successfully fixed to use an ESP as well: https://templates.blakadder.com/PA-210WYS.html

The bottom line right now is, "it is a smoke detector, but it is not smart (right now)". I did not test, but if I provide smoke to it, it will certainly start making noise, I'm pretty sure of that (but stopped smoking some time ago, so I need to really burn something to test)...

**Comment by sfromis** [comment]
Spying on serial communication while operating with factory firmware can be a good way of learning what happens.

**Comment by belveder79** [comment]
Ok, so I reassembled it with the original MCU and it seems that the MCU needs to receive a certain network status message proactively sent from the CB3S/ESP12  in order to start reporting. This message is not sent. Moreover I found that the MCU is not using the [standard serial communication protocol](https://developer.tuya.com/en/docs/iot/tuya-cloud-universal-serial-port-access-protocol?id=K9hhi0xxtn9cb), but rather uses the [Low-Power WIFI](https://developer.tuya.com/en/docs/iot/tuyacloudlowpoweruniversalserialaccessprotocol?id=K95afs9h4tjjh) one. I was able to make some sense out of the data (so dpid 23 is temperature, dpid 24 is humidity and 15 is battery), but I was not able yet to provide an initialization sequence from the ESP that worked.

for ex.
```lua
55 AA 00 05 00 08 17 02 00 04 00 00 01 17 41
```
header, 05 for report cmd, 08 for 8 bytes, dpid 17 so decimal 23, 02 for value, len is 0004 so 4 bytes, and the value is 01 17 which is 279 relating to 27.9 degrees Celsius, and checksum

Are there any thoughts on the development side of this? The problem seems to be that the current tasmota drivers are for the standard serial and not the low-power serial - I did not succeed in going forward and backward with flashing different custom versions from Gitpod, and that's tedious (old - minimal - new - repeat). 

My thoughts were on either configuring the ESP12f as standard module (disabling the TUYA driver for now) and do SerialSend5 with hex, but how to I read something back repeatedly from the MCU? The other thought was to replace the ESP12f with an ESP32 temporarily and create a berry script to figure out the right order of commands... any thoughts?

**Comment by belveder79** [comment]
Let me quickly share my findings so far. I first sniffed the communication between the CB3S and the MCU and verified that the protocol is indeed the one written down in the Low-Power description... then I used an ESP32 and Berry to successfully trigger the readout based on providing the right packets (and ACKs for each one).

And finally... I hacked the driver in the current master branch 12.5.0. 

[changes_230625.zip](https://github.com/arendst/Tasmota/files/11859422/changes_230625.zip)

So briefly:
- added a define ´´´USE_SMOKE_SENSOR´´´ to the my_user_config
- changed a lot in V1, mainly redefining the CMDs and acting passive without heartbeat, just acknowledging packets 
  - MCU powers ESP, waits... 
  - ESP: send query device -> MCU: replies type/version -> ESP: send NetworkOK -> MCU: acks
  - MCU: send dpid/sensor -> ESP: acks with standard message, repeats until done
  - MCU is done, turns off ESP
- the V2 was just changed just in order to not complain about changes in V1

So far it pushes MQTT messages for the temperature/humidity/battery (dpid 23,24,15). What is what was based on guessing and comparing with plausible values. 

What I did not do yet is "put it on fire". It acts as a passive sensor right now. I also did not check what happens in case of an alarm, but that is what I'm going to do next. I will provide a modified version of the existing protocol to work at least for this sensor. I'll see if I can get a few more of such devices into my hands (e.g. door sensor) to expand the changes in the future. I would need someone from the team to propose a plausible way to make that permanently integrated, however...

**Comment by belveder79** [comment]
More fun... I tried heating up the board with a hairdryer, but I couldn't get it to sound the alarm - it reports temperature/humidity however. Probably the alarm is only triggered when there is heat AND smoke. Going to provide both soon...

Anyway, fun fact: once you reach 42° Celsius (around that) 0x000001A4, the payload of the temperature data reported flips to something like 0xFFFFFEDDF or so and then counts upwards from those negative values- so from -124° to -80° and so on as it gets warmer and warmer... it is very weird and does not make sense to me at all...

**Comment by belveder79** [comment]
Heureka!

Ok, so I also got some other ones, actually [those](https://de.aliexpress.com/item/1005004930330522.html?spm=a2g0o.order_list.order_list_main.35.21ef5c5fzmORrP&gatewayAdapt=glo2deu) ones. I think they are a little duller than the ones with the temp/humidity sensors. However, they are same platform - CB3S. I captured the traffic and modified the actual sources, adding ESP12F instead of CB3S. I have now a version which reports everything through MQTT. 

Here are a few things I learned for this one:
- it will report Alarm set to "On" if it detects smoke, and be offline all the time.
- it occasionally reports Battery status and Sensitivity status (I actually changed the sources to report anyway. The standard code only reports changes, but given that it goes on so rarely, who cares?)
- you cannot turn off the alarm, you can only turn off the annoying beeping sound. If you turn it off (```TuyaSend1 16,1``` in my code), it will power off the ESP. The MCU will stay on as long as it detects smoke; however, it will power the ESP again and report Alarm set to "Off" once it is done.
- you need to reroute the function ID ```TuyaMCU 11,2``` before starting, because per default it is set to dpid = 1 and that is the dpid the alarm is reported. In the end, the webserver toggle does nothing, but it does not matter as it is offline in 99.99% of cases anyway.
- It has a sensitivity setting, but this is "queried from the cloud" every time the ESP is powered. You can only reply to the query packet with whatever you want as a setting - for now I reply "no news", which keeps the sensitivity level to the default = low. I'd need to flash an other version with medium or high settings to change this.
- I intentionally left out every Software Update, WIFI reconfigure or other things the MCU might request, as I have not seen such messages at all and I wanted to keep it simple (Network connectivity is always there, so to say). 

I will put things together and let it report over my Node-Red + Telegram connection for a couple of days to see if this works (with some occasional induced fire alarms). Then I will move on with the other one with humidity/temperature.

---

So I'm done with this one, the other ones will follow and then there are a few door sensors I expect to behave the same. Now here is the question for you folks. The protocol is really different in large parts, including the Defines for the individual commands and messages, and the way you communicate with the module is different from standard, as the MCU is really taking the lead in controlling the ESP. There is nothing like fnIds basically, as there is only a "react to some action of the MCU", but no control and basically no access to the ESP at all, unless you trigger an alarm and let it buzz for minutes... How should that be dealt with in the future? Any ideas?
 
<img width="1636" alt="smokey_protocol" src="https://github.com/arendst/Tasmota/assets/112880/18f548c5-5bbb-41c6-ac28-c72d474eaad1">



[changes_230702.zip](https://github.com/arendst/Tasmota/files/11931269/changes_230702.zip)
![smokey_nodered](https://github.com/arendst/Tasmota/assets/112880/506bc117-6ce1-403e-8dd0-4f161eb66605)

**Comment by belveder79** [comment]
Ok, so I figured out that there might be an issue with the reconnect/wakeup of the ESPs. 

Since they tend to wake up really rarely, it might take quite a while to get the ESP registered to the WIFI network. This means that my approach, i.e. telling the MCU that "everything is fine" anyway, is a bad idea, because this will make the MCU send its data even if the network is not ready yet and the MQTT messages will not get delivered.

The other aspect is the unflashability of the devices. For a dumb smoke detector it might be fine to be flashed once and that's it, but thinking about it again I can't even update the firmware now without opening, because the MCU will pull power from the device always.

Is there a way to actually write a "protocol" based on Rules only? I was wondering if I could e.g. create a rule like MQTTConnected->TuyaSend some binary data, then a rule to do something on TuyaReceived XYZ... I read something about this, but I am not really fit with rules...

Will continue playing around with it and provide updates...

**Comment by belveder79** [comment]
Update:

[changes_230712.zip](https://github.com/arendst/Tasmota/files/12025465/changes_230712.zip)

- now sends network status over custom new ```TuyaSend10-TuyaSend12``` commands
- now allows setting the sensitivity over ```TuyaSend13-TuyaSend15``` commands
- uses rules


````
TuyaMCU 11,2
Rule1 ON Wifi#Connected DO Backlog Delay 50; TuyaSend11 ENDON
Rule2 ON Mqtt#Connected DO Backlog Delay 50; TuyaSend12 ENDON
Rule3 ON TuyaReceived#Data=55AA001000010010 DO Backlog Delay 10; TuyaSend14 ENDON
Rule1 ON
Rule2 ON
Rule3 ON
````

I suspect I can also change Rule3 to be communicated to the ESP upon connection, such that it acts like the Tuya Cloud - you change it somewhere and retain an update over mqtt, and the ESP updates the Rule upon an occasional connection... any help with this appreciated!
